APE14.rst

```diff
@@ -269,21 +284,42 @@ low-level API recommends:
             """
             A dictionary with each key being a string key from
             ``world_axis_object_components``, and each value being a tuple with
-            two elements:
+            two or three elements:
 
             * The first element of the tuple must be a string specifying the
               fully-qualified name of a class, which will specify the actual
               Python object to be created.
 
-            * The second tuple element must be a
-              dictionary with the keyword arguments required to initialize the
-              class.
+            * The second element, which is optional, should be a tuple
+              specifying the positional arguments required to initialize the
+              class. If ``world_axis_object_components`` specifies that
+              the world coordinates should be passed as a positional argument,
+              this this tuple should include ``None`` placeholders for the
+              world coordinates. This tuple only needs to be specified if there
+              are additional positional arguments.
+
+            * The last tuple element must be a dictionary with the keyword
+              arguments required to initialize the class.
 
             See below for an example of this property. Note that we don't
             require the classes to be Astropy classes since there is no
             guarantee that Astropy will have all the classes to represent all
             kinds of world coordinates. Furthermore, we recommend that the
             output be kept as human-readable as possible.
+
+            The classes used here should have the ability to do conversions by
+            passing an instance as the first argument to the same class with
+            different arguments (e.g. ``Time(Time(...), scale='tai')``). This is
+            a requirement for the implementation of the high-level interface.
+
+            The second tuple element for each value of this dictionary can in
+            turn contain either instances of classes, or if necessary can contain
+            serialized versions that should take the same form as the main
+            classes described above (a tuple with three elements with the
+            fully qualified name of the class, then the positional arguments
+            and the keyword arguments). For low-level API objects implemented
+            in Python, we recommend simply returning the actual objects (not
+            the serialized form) for optimal performance.
```

Comment:
One open question - how do we know whether a tuple is a serialized version or the actual Python object?

Reply:
Why not just mandate realised python classes here? What is this addition trying to solve?

Reply:
The API was designed so that if one wanted to implement it in C with the Python API, that could be done. For that to work, we have to allow objects to be serialized. However, we then realized this would have a performance hit when it's not needed, so this addition was to say that it's ok to just use Python classes. Maybe this should be rephrased to say that this is the standard, and then to give the serialization as an option.

Reply:
> For that to work, we have to allow objects to be serialized.

I mean *have to* is strong here, it's perfectly possible to create Python objects from C. I think we need to walk this line carefully.

Also is it likely that a tuple will actually ever be the desired Python object?

Reply:
This was addressed in https://github.com/astropy/astropy-APEs/pull/45/commits/8196354329cc2f7c671007dc94781392e3f097cb following a discussion on a telecon.